### PR TITLE
refactor: added a "common" package to share code between packages

### PR DESCRIPTION
**Pull request by andreabadesso**
### Motivation

We should have a `common` package so that the wallet-service and the daemon can share code between them.

This PR does not attempt to refactor both projects, but only to create the common project and include the bare minimum changes to enable code sharing of the nft util

### Acceptance Criteria

- We should migrate the `addAlert` from the daemon's package to the common package and use it on both the wallet-service and the daemon
- We should migrate types related to transactions and use them in the daemon

### Checklist
- [X] If you are requesting a merge into `master`, confirm this code is production-ready and can be included in future releases as soon as it gets merged
- [X] Make sure either the unit tests and/or the QA tests are capable of testing the new features
- [X] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,22 +3,40 @@
...
+  "dependencies": {
+    "@aws-sdk/client-apigatewaymanagementapi": "3.540.0",
+    "@aws-sdk/client-lambda": "3.540.0",
+    "@aws-sdk/client-sqs": "3.540.0",
+    "@hathor/wallet-lib": "0.39.0",
+    "bip32": "^4.0.0",
+    "bitcoinjs-lib": "^6.1.5",
+    "bitcoinjs-message": "^2.2.0",
+    "tiny-secp256k1": "^2.2.3",
+    "winston": "^3.13.0"
```
Comment (andreabadesso) [inline package.json]: Added those dependencies here so the projects can use `peerDependencies` and share the dependency

File: packages/common/src/utils/index.utils.ts
Comment (tuliomir) [inline packages/common/src/utils/index.utils.ts:1]: chore: We should have the Hathor header on this file

**Review by tuliomir — APPROVED** [review]

File: packages/common/src/utils/index.utils.ts
Comment (andreabadesso) [reply packages/common/src/utils/index.utils.ts:1]: Fixed in 497f843638571767f283d3cc58034173aa589de9

**Review by tuliomir — APPROVED** [review]

File: packages/common/src/utils/alerting.utils.ts [hunk 2]
```diff
@@ -28,9 +28,9 @@ export const addAlert = async (
   title: string,
   message: string,
   severity: Severity,
-  metadata?: unknown,
+  metadata: unknown = null,
```
Comment (pedroferreira1) [inline packages/common/src/utils/alerting.utils.ts]: Why did you set this default value? Shouldn't we have the optional parameters in the end of the method?

File: packages/daemon/package.json [hunk 3]
```diff
@@ -32,20 +32,26 @@
     "@types/mysql": "^2.15.21",
     "@types/node": "^17.0.45",
     "@types/ws": "^8.5.5",
-    "@typescript-eslint/eslint-plugin": "^6.7.3",
-    "@typescript-eslint/parser": "^6.7.3",
+    "@typescript-eslint/eslint-plugin": "^7.4.0",
```
Comment (pedroferreira1) [inline packages/daemon/package.json]: Why did you upgrade this major version of this package? Shouldn't we do any upgrades in a separate PR to be better reviewed?

File: packages/daemon/package.json [hunk 4]
```diff
@@ -32,20 +32,26 @@
...
+    "@typescript-eslint/parser": "^7.4.0",
+    "eslint": "^8.57.0",
```
Comment (pedroferreira1) [inline packages/daemon/package.json]: Why do you need `eslint` now in the `daemon`?

File: packages/wallet-service/package.json [hunk 5]
```diff
@@ -36,22 +33,21 @@
...
+    "uuid": "^8.3.0"
   },
   "devDependencies": {
     "@types/aws-lambda": "^8.10.95",
     "@types/jest": "^27.0.24",
     "@types/node": "^18.0.4",
-    "@typescript-eslint/eslint-plugin": "^6.7.4",
-    "@typescript-eslint/parser": "^3.3.0",
+    "@typescript-eslint/eslint-plugin": "^7.4.0",
+    "@typescript-eslint/parser": "^7.4.0",
```
Comment (pedroferreira1) [inline packages/wallet-service/package.json]: Why did you upgrade this lib? It was 4 majors before.

File: packages/wallet-service/package.json [hunk 6]
```diff
@@ -72,5 +68,12 @@
     "typescript-eslint": "0.0.1-alpha.0",
     "webpack": "^5.88.2",
     "webpack-node-externals": "^3.0.0"
+  },
+  "peerDependencies": {
+    "@aws-sdk/client-apigatewaymanagementapi": "3.540.0",
+    "@aws-sdk/client-lambda": "3.540.0",
+    "@aws-sdk/client-sqs": "3.540.0",
+    "@hathor/wallet-lib": "^0.39.0",
+    "winston": "^3.13.0"
```
Comment (pedroferreira1) [inline packages/wallet-service/package.json]: Any reason to upgrade this package?

File: packages/wallet-service/src/types.ts [hunk 7]
```diff
@@ -798,15 +751,4 @@ export interface WalletBalanceValue {
...
- */
-export enum Severity {
-  CRITICAL = 'critical',
-  MAJOR = 'major',
-  MEDIUM = 'medium',
-  MINOR = 'minor',
-  WARNING = 'warning',
-  INFO = 'info',
-}
+export { Severity, Transaction, TxInput, TxOutput, DecodedOutput };
```
Comment (pedroferreira1) [inline packages/wallet-service/src/types.ts]: Why do you need to export the imported types?

File: package.json [hunk 8]
```diff
@@ -3,22 +3,40 @@
...
-  "packageManager": "yarn@4.1.0"
+  "packageManager": "yarn@4.1.0",
+  "dependencies": {
+    "@aws-sdk/client-apigatewaymanagementapi": "3.540.0",
+    "@aws-sdk/client-lambda": "3.540.0",
+    "@aws-sdk/client-sqs": "3.540.0",
+    "@hathor/wallet-lib": "0.39.0",
+    "bip32": "^4.0.0",
+    "bitcoinjs-lib": "^6.1.5",
+    "bitcoinjs-message": "^2.2.0",
```
Comment (pedroferreira1) [inline package.json:40]: Does it make sense to have this as peerDependency in the wallet service package.json?  You are duplicating this package

File: packages/daemon/package.json (hunk 4 above)
Comment (andreabadesso) [reply packages/daemon/package.json]: Reverted all package changes related to eslint, I'll do it in this PR:
https://github.com/HathorNetwork/hathor-wallet-service/pull/156

File: packages/daemon/package.json (hunk 3 above)
Comment (andreabadesso) [reply packages/daemon/package.json]: Reverted all package changes related to eslint, I'll do it in this PR:
https://github.com/HathorNetwork/hathor-wallet-service/pull/156

File: packages/wallet-service/package.json (hunk 5 above)
Comment (andreabadesso) [reply packages/wallet-service/package.json]: Reverted all package changes related to eslint, I'll do it in this PR:
https://github.com/HathorNetwork/hathor-wallet-service/pull/156

File: packages/wallet-service/package.json (hunk 6 above)
Comment (andreabadesso) [reply packages/wallet-service/package.json]: Since we were using caret ranges in the version (`^3.3.0`), we were already installing the latest non-major version (< `4.0.0`) after a `yarn install`

image

image

So this will not cause the package to get updated, it's just setting the version we're actually using

File: packages/wallet-service/src/types.ts (hunk 7 above)
Comment (andreabadesso) [reply packages/wallet-service/src/types.ts]: We don't, this was done initially because I was only planning to move the types on the next PR

But refactored the code to import from the common project, can you please check again?

File: package.json (hunk 8 above)
Comment (andreabadesso) [reply package.json:40]: It does, refactored in b6e4368e865b8947ab22ab329343da05891117c5

File: packages/common/src/utils/alerting.utils.ts (hunk 2 above)
Comment (andreabadesso) [reply packages/common/src/utils/alerting.utils.ts]: Added a comment in this line: a410a8ca63de34d48327ca4e735cc2ae62f03d12

I'll refactor the `logger` to be shared between projects, so `logger` will stop being a parameter there, so this default to null is temporary until then

Comment (pedroferreira1) [reply packages/common/src/utils/alerting.utils.ts]: I still don't understand the default `null`. We use default values for parameters in the end of the list of parameters, which is not the case.

File: packages/wallet-service/tests/integration.test.ts [hunk 9]
```diff
@@ -4,7 +4,8 @@ import eventTemplate from '@events/eventTemplate.json';
 import { loadWallet, loadWalletFailed } from '@src/api/wallet';
 import { createWallet, getMinersList } from '@src/db';
 import * as txProcessor from '@src/txProcessor';
-import { Transaction, WalletStatus, TxInput, Severity } from '@src/types';
+import { WalletStatus  } from '@src/types';
```
Comment (tuliomir) [inline packages/wallet-service/tests/integration.test.ts]: style: Lint fix
```suggestion
import { WalletStatus } from '@src/types';
```

File: packages/wallet-service/tests/utils/pushnotification.utils.test.ts [hunk 10]
```diff
@@ -6,7 +6,8 @@ import { sendMulticastMock, messaging, initFirebaseAdminMock } from '@tests/util
 import { logger } from '@tests/winston.mock';
 import { PushNotificationUtils, PushNotificationError, buildFunctionName, FunctionName } from '@src/utils/pushnotification.utils';
 import * as pushnotificationUtils from '@src/utils/pushnotification.utils';
-import { SendNotificationToDevice, Severity } from '@src/types';
+import { SendNotificationToDevice } from '@src/types';
+import { Severity  } from '@wallet-service/common/src/types';
```
Comment (tuliomir) [inline packages/wallet-service/tests/utils/pushnotification.utils.test.ts]: style: Lint fix
```suggestion
import { Severity } from '@wallet-service/common/src/types';
```

File: packages/common/package.json [hunk 11]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "@wallet-service/common",
+  "packageManager": "yarn@4.1.0",
+  "peerDependencies": {
+    "@aws-sdk/client-lambda": "3.540.0",
+    "@hathor/wallet-lib": "0.39.0",
+    "winston": "^3.13.0"
```
Comment (tuliomir) [inline packages/common/package.json:11]: question(non-blocking): Here we have a mix of version definitions. Some are pinned to an exact value and some are with a flexible caret range.

Should we decide for one approach or another? Or can we leave this mixed as it currently is?

File: package.json [hunk 12]
```diff
@@ -3,22 +3,40 @@
   "version": "1.5.0",
   "workspaces": [
     "packages/daemon",
-    "packages/wallet-service"
+    "packages/wallet-service",
+    "packages/common"
```
Comment (tuliomir) [inline package.json]: suggestion: Order this alphabetically, as it will probably be automatically fixed at some point in the future
```js,
    "packages/common",
    "packages/daemon",
    "packages/wallet-service"
```

**Review by tuliomir — APPROVED** [review]

File: packages/common/package.json (hunk 11 above)
Comment (andreabadesso) [reply packages/common/package.json:11]: I'll fix all versions in a future PR

File: package.json (hunk 12 above)
Comment (andreabadesso) [reply package.json]: Done in 408d995a493a80f4ca8d17e1afe10bf8c669f2e1

File: packages/wallet-service/tests/utils/pushnotification.utils.test.ts (hunk 10 above)
Comment (andreabadesso) [reply packages/wallet-service/tests/utils/pushnotification.utils.test.ts]: Done in 408d995a493a80f4ca8d17e1afe10bf8c669f2e1

File: packages/wallet-service/tests/integration.test.ts (hunk 9 above)
Comment (andreabadesso) [reply packages/wallet-service/tests/integration.test.ts]: Done in 408d995a493a80f4ca8d17e1afe10bf8c669f2e1

File: packages/common/src/utils/alerting.utils.ts (hunk 2 above)
Comment (andreabadesso) [reply packages/common/src/utils/alerting.utils.ts]: Removed the null default in cae81acc04b833d561559f1f5f8156481057c40d

**Review by pedroferreira1 — APPROVED** [review]
